cfep-09.md

```diff
@@ -0,0 +1,280 @@
+
+<table>
+<tr><td> Pinning Proposal System for Automatic Rebuilds</td><td> Pinning Proposal </td>
+<tr><td> Status </td><td> Proposed </td></tr>
+<tr><td> Author(s) </td><td> Christopher J. "CJ" Wright &lt;cjwright4242@gmail.com&gt;; Matthew R. Becker &lt;becker.mr@gmail.com&gt;</td></tr>
+<tr><td> Created </td><td> Feb 20, 2019</td></tr>
+<tr><td> Updated </td><td> Aug 30, 2016</td></tr>
+<tr><td> Discussion </td><td> NA </td></tr>
+<tr><td> Implementation </td><td> NA </td></tr>
+</table>
+
+## Abstract
+
+This CFEP proposes a change to how our internal pinning system works.
+The new system will comprise of a new repo for proposed migrations.
+To propose a migration a PR will be issued with a new version of
+the global pinning file and some metadata about the migration needed 
+(if any).
+Once the PR is squashed and merged the bot will take the metadata
+to calculate which packages need to be rebuilt and start issuing
+PRs with the new pinning as a local pinning file to the feedstocks
+in topo order.
+
+This CFEP is taken from @beckermr's work on [conda-forge.github.io](https://github.com/conda-forge/conda-forge.github.io/issues/712) and @@mariusvniekerk's work in [conda-smithy](https://github.com/conda-forge/conda-forge.github.io/issues/712).
+
+## Motivation
+Currently when we move pinnings we move the pin in the ``conda-forge-pinning-feedstock``.
+Then we start a migration using the ``regro-cf-autotick-bot``.
+This order of operations causes some issues, when new feedstocks
+are created they receive the new pinnings even if their dependencies
+are not rebuilt with the new pinnings themselves.
+
+## Specification
+
+### The migrations proceed as follows:
+1. We make a PR into the ``migrations`` folder in the ``conda-forge-pinning-feedstock`` with a new yaml file representing the migration
+2. Once the change is merged, the bot picks it up, builds a migrator object and begins the migration process
+3. A migration PR is issued for a node only if 
+    1. The node depends on the changed pinnings.
+    2. It has no dependencies that depend on the new pinnings and have not been migrated.
+4. Process 3 continues until we determine that the migration is sufficiently complete and the change is applied to the global pinning file via a PR.
+
+### Implementation Details:
+Given a feedstock with the following pinnings
+
+```yaml
+# Existing pins
+a:
+  - 1.5
+b:
+  - 1.2
+someflag:
+  - disabled
+```
+
+There are four classes of pinning we can do in general.
+
+#### 1. Simple Additive Version Migrations
```

Comment:
Does this syntax support bumping pins for only one platform? 

For example

```yaml
# migrator
__migrator: kind: version
a:
  - 1.5  # [not osx]
  - 1.6   # [osx]
b:
  - 1.1
```

Reply:
@mariusvniekerk

Reply:
I think this will work. Since the rendering (which handles the selectors) will happen first.